sycl/include/sycl/nd_item.hpp

```diff
@@ -38,6 +39,273 @@ namespace ext::oneapi::experimental {
+
+    return group<Dimensions>::async_work_group_copy(dest, src, numElements,
+                                                    srcStride);
+  }
+
+  template <typename dataT>
+  __SYCL2020_DEPRECATED("Use decorated multi_ptr arguments instead")
+  device_event
+      async_work_group_copy(global_ptr<dataT> dest, local_ptr<dataT> src,
+                            size_t numElements, size_t destStride) const {
+    return group<Dimensions>::async_work_group_copy(dest, src, numElements,
+                                                    destStride);
+  }
+
+  template <typename DestDataT, typename SrcDataT>
+  typename std::enable_if_t<
+      std::is_same_v<DestDataT, std::remove_const_t<SrcDataT>>, device_event>
+  async_work_group_copy(decorated_local_ptr<DestDataT> dest,
+                        decorated_global_ptr<SrcDataT> src,
+                        size_t numElements) const {
+    return group<Dimensions>::async_work_group_copy(dest, src, numElements);
+  }
+
+  template <typename DestDataT, typename SrcDataT>
+  typename std::enable_if_t<
+      std::is_same_v<DestDataT, std::remove_const_t<SrcDataT>>, device_event>
+  async_work_group_copy(decorated_global_ptr<DestDataT> dest,
+                        decorated_local_ptr<SrcDataT> src,
+                        size_t numElements) const {
+    return group<Dimensions>::async_work_group_copy(dest, src, numElements);
+  }
+
+  template <typename DestDataT, typename SrcDataT>
+  typename std::enable_if_t<
+      std::is_same_v<DestDataT, std::remove_const_t<SrcDataT>>, device_event>
+  async_work_group_copy(decorated_local_ptr<DestDataT> dest,
+                        decorated_global_ptr<SrcDataT> src, size_t numElements,
+                        size_t srcStride) const {
+
+    return group<Dimensions>::async_work_group_copy(dest, src, numElements,
+                                                    srcStride);
+  }
+
+  template <typename DestDataT, typename SrcDataT>
+  typename std::enable_if_t<
+      std::is_same_v<DestDataT, std::remove_const_t<SrcDataT>>, device_event>
+  async_work_group_copy(decorated_global_ptr<DestDataT> dest,
+                        decorated_local_ptr<SrcDataT> src, size_t numElements,
+                        size_t destStride) const {
+    return group<Dimensions>::async_work_group_copy(dest, src, numElements,
+                                                    destStride);
+  }
+
+  template <typename... eventTN> void wait_for(eventTN... events) const {
+    group<Dimensions>::wait_for(events...);
+  }
+
+  sycl::ext::oneapi::experimental::root_group<Dimensions>
+  ext_oneapi_get_root_group() const {
+    return sycl::ext::oneapi::experimental::root_group<Dimensions>{*this};
+  }
+
+  nd_item(const nd_item &rhs) = default;
+  nd_item(nd_item &&rhs) = default;
+
+  nd_item &operator=(const nd_item &rhs) = default;
+  nd_item &operator=(nd_item &&rhs) = default;
+
+  bool operator==(const nd_item &) const { return true; }
```

Comment:
This is an important question because it can kill the whole idea of nd_item optimization.
My initial thought was that `nd_item` object can be created only by the runtime and it's not modifiable. In other words, we can consider it to be singleton object for the work-item.
On the other hand, I can see that `operator==` might need to return `false` if `nd_item` argument is created by another kernel instance (i.e. work-item). Now the question is: can we get access to `nd_item` object from another work-item?
I can imagine that the only option is to pass it though the memory shared among work-items (i.e. `local` or `global`), but I can't come-up with the example where it can be done w/o adding UB.
Is it possible to allocate a shared memory for nd_item object and store/load this object using only a copy constructor?

Reply:
> but I can't come-up with the example where it can be done w/o adding UB.

I think this can be done by using placement-new and the copy constructor.  The application can allocate uninitialized bytes in shared memory and then use placement-new to copy-construct an `nd_item` object into that memory.

Once the object is in shared memory, it could be accessed by other work-items in the kernel execution.

Reply:
> The application can allocate uninitialized bytes in shared memory and then use placement-new to copy-construct an `nd_item` object into that memory.

Does this involve type casting (from "uninitialized bytes" to "nd_item")? If so, is this casting properly specified? At least "uninitialized bytes" must be properly aligned.

Reply:
Placement new is a standard C++ feature, and it does not require any type casting.  It is true that the "uninitialized bytes" need to be properly aligned.  You can see more in the [cppreference description](https://en.cppreference.com/w/cpp/language/new#Placement_new).

I admit this is an unusual code pattern, and I don't know if it is something that any real application does.  However, I think it is legal according to the current wording of the spec.

Reply:
I think this is something that we need to revisit and clarify in the SYCL specification.

Between SYCL 1.2.1 and SYCL 2020, the definition of the `group` class changed to make it clear that `group` (and `sub_group`) are stateless.  I'm having trouble tracking down the commit (and I suspect it's internal to Khronos anyway) but you can see that the 2020 definitions say things like "Return a SYCL `id` representing **the calling work-item’s position** within the `work-group`" (emphasis mine).

Does `nd_item` only have an `operator==` because of Section 4.5.3?  This whole section seems weird to me, now: firstly, because most of these classes are difficult to pass between work-items; and secondly, because the equality operator seems underspecified.  The specification says:

> Equality between two instances of `T` (i.e. `a == b`) must be true if the value of all members are equal

...but the specification for `nd_item` doesn't say what the members are, or if there even are any.

Reply:
> Does `nd_item` only have an `operator==` because of Section 4.5.3? This whole section seems weird to me, now: firstly, because most of these classes are difficult to pass between work-items; and secondly, because the equality operator seems underspecified. The specification says:
> 
> > Equality between two instances of `T` (i.e. `a == b`) must be true if the value of all members are equal
> 
> ...but the specification for `nd_item` doesn't say what the members are, or if there even are any.

That's exactly right. I think SYCL spec should allow "member-less" implementation of `nd_item` like I tried to prototype in this PR. 
I'm almost 100% sure that no users who passes `nd_item` objects between work-items, but Greg might be right that there is a loophole allowing this use case technically even if has little sense in practice.

Reply:
>I think this is something that we need to revisit and clarify in the SYCL specification

I created this issue in the SYCL spec repo: KhronosGroup/SYCL-Docs#532.

Reply:
Following up on this discussion: if the implementation is "member-less", does it make sense to have equality/inequality comparison at all? I don't feel very comfortable with a `==` method that always returns `true`, but I can live with it.

Reply:
With the current wording of the spec, that class has to have `operator==`, because it follows common by-value semantics. I would expect that if we agree that `nd_item` represents a calling work-item, then it doesn't make a lot of sense to have `operator==` in it and therefore the spec will be updated to reflect that.

Reply:
I'll do follow-up updates if spec changes. Meanwhile, let's keep the operator.